README.md
# PSSyncthing

Allows you to manage Syncthing from PowerShell



CHANGELOG.md
## v0.0.1 Initial Release


source/public/Get-SyncthingConfig.ps1
function Get-SyncthingConfig {
    <#
    .SYNOPSIS
        Get's the configuration of Syncthing.
    .DESCRIPTION
        Get's the configuration of Syncthing on local or remote computers.
    .EXAMPLE
        Get-SyncthingConfig -ApiKey 'abc123'

        Returns the configuration of Syncthing on the local computer.
    .EXAMPLE
        Get-SyncthingConfig -ComputerName 'server01' -Port 8090 -ApiKey 'abc' -ValidateCertificate

        Get's the Syncthing configuration from computer server01 on port 8090 using api key 'abc' and validates the Syncthing SSL certificate. 
    .OUTPUTS
        [PSCustomObject] containing the configuration if we can connect to
        Syncthing. Otherwise we throw an exception.
    .NOTES
        Author  : Paul Broadwith (https://github.com/pauby)
        Project : PSSyncthing (https://github.com/pauby/PSSyncthing)
        History : 1.0 - 07/05/18 - Initial release
    #>

    [OutputType([PSCustomObject])]
    [CmdletBinding()]
    Param (
        # ComputerName to connect to that has Syncthing running. Defaults to 'localhost'.
        [Parameter(ValueFromPipeline, ValueFromPipelineByPropertyName)]
        [string[]]
        $ComputerName = 'localhost',

        # Port Syncthing is listening on. Defaults to '8384'.
        [Parameter(ValueFromPipelineByPropertyName)]
        [ValidateRange(1, 65535)]
        [int]
        $Port = 8384,

        # Syncthing API Key to use to connect.
        [Parameter(Mandatory, ValueFromPipelineByPropertyName)]
        [string]
        $ApiKey
    )
    
    Begin {
        $params = @{
            Port                = $Port
            ApiKey              = $ApiKey
            Endpoint            = '/system/config'
        }
    }

    Process {
        ForEach ($name in $ComputerName) {
            Write-Verbose "Getting Syncthing configuration from '$name'."
            $params.ComputerName = $name

            $response = Invoke-SyncthingRequest @params
            if ($null -eq $response) {
                return $null
            }
            else {
                return $response
            }
        }
    }
}

source/public/Test-SyncthingConfigInSync.ps1
function Test-SyncthingConfigInSync {
    <#
    .SYNOPSIS
        Tests whether Syncthing configuration is in sync.
    .DESCRIPTION
        Tests whther Syncthings the running configuration is in sync with the
        configuration on disk.
    .EXAMPLE
        Test-SyncthingConfigInSync -ApiKey 'abc123'

        Tests if the running configuration of Syncthing on the local computer using api key abc123 is the same as that on disk, or $false is not.
    .EXAMPLE
        Test-SyncthingConfigInSync -ComputerName 'server01' -Port 8090 -ApiKey 'abc' -ValidateCertificate

        Tests if the running configuration of Syncthing on computer server01, port 8090 using api key 'abc' and validating the Syncthing SSL certificate is the same as the configuration on disk.
    .OUTPUTS
        [Boolean] of $true if the running configuration is the same as that on disk and $false otherwise.
    .NOTES
        Author  : Paul Broadwith (https://github.com/pauby)
        Project : PSSyncthing (https://github.com/pauby/PSSyncthing)
        History : 1.0 - 07/05/18 - Initial release
    #>

    [OutputType([Boolean])]
    [CmdletBinding()]
    Param (
        # ComputerName to connect to that has Syncthing running. Defaults to 'localhost'.
        [Parameter(ValueFromPipeline, ValueFromPipelineByPropertyName)]
        [string[]]
        $ComputerName = 'localhost',

        # Port Syncthing is listening on. Defaults to '8384'.
        [Parameter(ValueFromPipelineByPropertyName)]
        [ValidateRange(1, 65535)]
        [int]
        $Port = 8384,

        # Syncthing API Key to use to connect.
        [Parameter(Mandatory, ValueFromPipelineByPropertyName)]
        [string]
        $ApiKey
    )

    Begin {        
        $params = @{
            Port                = $Port
            ApiKey              = $ApiKey
            Endpoint            = '/system/config/insync'
        }
    }

    Process {
        ForEach ($name in $ComputerName) {
            Write-Verbose "Checking configuration sync status of '$name'."
            $params.ComputerName = $name
            (Invoke-SyncthingRequest @params).configInSync
        }
    }

}

source/public/Enable-SyncthingCertValidation.ps1
function Enable-SyncthingCertValidation {
    <#
    .SYNOPSIS
        Enables SSL certificate validation.
    .DESCRIPTION
        Enables SSL certificate validation for the session. Note that this is
        not limited to SSL validation for Syncthing but for the entire session.
    .EXAMPLE
        Enable-SyncthingCertValidation

        Enables SSL certificate validation for the session.
    .NOTES
        Author  : Paul Broadwith (https://github.com/pauby)
        Project : PSSyncthing (https://github.com/pauby/pssyncthing)
        History : 1.0 - 07/05/18 - Initial release
    .LINK
        Disable-SyncthingCertificateValidation
    #>

    [CmdletBinding(SupportsShouldProcess, ConfirmImpact = 'Medium')]
    Param ()

    #TODO This needs to be written - code left here deliberately but this
    # disables the validation
    if ($PSCmdlet.ShouldProcess('SSL Certificate Validation', 'Disable')) {
        # if we do not ignore certificate validation then self-signed
        # certificates will throw up errors - this code ignore the certificate
        # validation - taken from
        # https://blog.ukotic.net/2017/08/15/could-not-establish-trust-relationship-for-the-ssltls-invoke-webrequest/
        if (-not ([System.Management.Automation.PSTypeName]'ServerCertificateValidationCallback').Type) {
            $certCallback = @"
using System;
using System.Net;
using System.Net.Security;
using System.Security.Cryptography.X509Certificates;
public class ServerCertificateValidationCallback
{
    public static void Ignore()
    {
        if(ServicePointManager.ServerCertificateValidationCallback ==null)
        {
            ServicePointManager.ServerCertificateValidationCallback += 
                delegate
                (
                    Object obj, 
                    X509Certificate certificate, 
                    X509Chain chain, 
                    SslPolicyErrors errors
                )
                {
                    return true;
                };
        }
    }
}
"@
            Add-Type $certCallback
        }
        [ServerCertificateValidationCallback]::Ignore()
    }
}

source/public/Disable-SyncthingCertValidation.ps1
function Disable-SyncthingCertValidation {
    <#
    .SYNOPSIS
        Disables SSL certificate validation.
    .DESCRIPTION
        Disables SSL certificate validation for the session. Note that this is
        not limited to SSL validation for Syncthing but for the entire session.
    .EXAMPLE
        Disable-SyncthingCertValidation

        Disables SSL certificate validation for the session.
    .NOTES
        Author  : Paul Broadwith (https://github.com/pauby)
        Project : PSSyncthing (https://github.com/pauby/pssyncthing)
        History : 1.0 - 07/05/18 - Initial release
    .LINK
        Enable-SyncthingCertificateValidation
    #>

    [CmdletBinding(SupportsShouldProcess, ConfirmImpact = 'Medium')]
    Param ()

    if ($PSCmdlet.ShouldProcess('SSL Certificate Validation', 'Disable')) {
        # if we do not ignore certificate validation then self-signed
        # certificates will throw up errors - this code ignore the certificate
        # validation - taken from
        # https://blog.ukotic.net/2017/08/15/could-not-establish-trust-relationship-for-the-ssltls-invoke-webrequest/
        if (-not ([System.Management.Automation.PSTypeName]'ServerCertificateValidationCallback').Type) {
            $certCallback = @"
using System;
using System.Net;
using System.Net.Security;
using System.Security.Cryptography.X509Certificates;
public class ServerCertificateValidationCallback
{
    public static void Ignore()
    {
        if(ServicePointManager.ServerCertificateValidationCallback ==null)
        {
            ServicePointManager.ServerCertificateValidationCallback += 
                delegate
                (
                    Object obj, 
                    X509Certificate certificate, 
                    X509Chain chain, 
                    SslPolicyErrors errors
                )
                {
                    return true;
                };
        }
    }
}
"@
            Add-Type $certCallback
        }
        [ServerCertificateValidationCallback]::Ignore()
    }
}

source/public/Invoke-SyncthingRequest.ps1
function Invoke-SyncthingRequest {
    <#
    .SYNOPSIS
        Invokes a REST request to Syncthing.
    .DESCRIPTION
        Invokes a REST request to Syncthing running on a computer.

        If you need to disable Syncthing SSL certificate validation use the Diable-SyncthingCertValidation
    .EXAMPLE
        Invoke-SyncthingRequest -ApiKey 'abc123' -EndPoint '/system/browse'

        Invoke a REST request to localhost on port 8384 at endpoint /rest/system/browse using the api key 'abc123'.
    .EXAMPLE
        Invoke-SyncthingRequest -ComputerName 'server01' -Port 8090 -ApiKey 'abc' -Endpoint '/system/config' -EndpointParameter @{ 'current' = 'var' }

        Invoke a REST request to server01 on port 8090 using api key 'abc' and connecting to the endpoint /system/config with a parameter of current=var. 
    .OUTPUTS
        [PSCustomObject]
    .NOTES
        Author  : Paul Broadwith (https://github.com/pauby)
        Project : PSSyncthing (https://github.com/pauby/PSSyncthing)
        History : 1.0 - 03/05/18 - Initial release
    #>

    [OutputType([PSCustomObject])]
    [CmdletBinding()]
    Param (
        # ComputerName to connect to that has Syncthing running. Defaults to 'localhost'.
        [Parameter(ValueFromPipelineByPropertyName)]
        [string[]]
        $ComputerName = 'localhost',

        # Port Syncthing is listening on. Defaults to '8384'.
        [Parameter(ValueFromPipelineByPropertyName)]
        [ValidateRange(1, 65535)]
        [int]
        $Port = 8384,

        # Syncthing API Key to use to connect.
        [Parameter(Mandatory, ValueFromPipelineByPropertyName)]
        [string]
        $ApiKey,

        [Parameter(ValueFromPipelineByPropertyName)]
        [ValidateSet('Get', 'Post')]
        [string]
        $Method = 'Get',

        # The REST endpoint to contact. Make sure to omit the intiial '/rest' -
        # so endpoint '/rest/system/browse' should just be '/system/browse'
        [Parameter(Mandatory, ValueFromPipelineByPropertyName)]
        [string]
        $Endpoint,

        # Parameters to be passed to the endpoint. This should be in the @{ key
        # = value } format (ie. a hashtable).
        [Parameter(ValueFromPipelineByPropertyName)]
        [hashtable]
        $EndpointParameter
    )

    # the Headers parameter needs a hashtable so lets create one with
    # Syncthing's API Key
    $header = @{ 'X-API-Key' = $ApiKey }
    $msg = $header | ForEach-Object { "$($_.keys) = $($_.values);" }
    Write-Verbose "Syncthing REST request header is: $msg"

    # lets build up the Uri
    $uri = "http://$($ComputerName):$($Port)/rest$Endpoint"

    # if there are any parameters then create the query string
    if ($EndpointParameter) {
        $uri += '?'
        $count = 0
        ForEach ($k in $EndpointParameter.Keys) {
            if ($count -gt 0) {
                $uri += '&'
            }

            $uri += "$k=$($EndpointParameter.$k)"
            $count++
        }
    }
    
    Write-Verbose "Invoking REST request to Syncthing on computer '$ComputerName', port '$Port' using '$Method' method to Uri '$uri'."
    $params = @{
        Method          = $Method
        Uri             = $uri
        Headers         = $header
        UseBasicParsing = $true
    }
    Invoke-RestMethod @params 
}